File: crates/bubbaloop-node-sdk/src/health.rs
use std::sync::Arc;
use tokio::sync::watch;

/// Spawn a background task that publishes health heartbeats every 5 seconds.
///
/// Publishes `"ok"` to `bubbaloop/{scope}/{machine_id}/health/{node_name}`.
/// Stops when the shutdown signal fires.
pub async fn spawn_health_heartbeat(
    session: Arc<zenoh::Session>,
    scope: &str,
    machine_id: &str,
    node_name: &str,
    mut shutdown_rx: watch::Receiver<()>,
) -> anyhow::Result<tokio::task::JoinHandle<()>> {
    let health_topic = format!("bubbaloop/{}/{}/health/{}", scope, machine_id, node_name);
    log::info!("Health heartbeat: {}", health_topic);
    let publisher = session
        .declare_publisher(health_topic)
        .await
        .map_err(|e| anyhow::anyhow!("Failed to create health publisher: {}", e))?;

    let handle = tokio::spawn(async move {
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(5));
        loop {
            tokio::select! {
                biased;
                _ = shutdown_rx.changed() => {
                    log::debug!("Health heartbeat stopping");
                    break;
                }
                _ = interval.tick() => {
                    if let Err(e) = publisher.put("ok").await {
                        log::warn!("Health heartbeat failed: {}", e);
                    }
                }
            }
        }
    });

    Ok(handle)
}

File: python-sdk/bubbaloop_sdk/health.py
"""Background health heartbeat thread."""

import threading
import time

import zenoh


def start_health_heartbeat(
    session: zenoh.Session,
    scope: str,
    machine_id: str,
    instance_name: str,
    shutdown: threading.Event,
    interval_secs: float = 5.0,
) -> threading.Thread:
    """Publish 'ok' to ``{instance_name}/health`` every ``interval_secs``.

    Returns the daemon thread (already started). Stops when ``shutdown`` is set.
    """
    topic = f"bubbaloop/{scope}/{machine_id}/{instance_name}/health"
    pub = session.declare_publisher(topic)

    def _loop():
        while not shutdown.wait(timeout=interval_secs):
            pub.put(b"ok")

    t = threading.Thread(target=_loop, daemon=True, name=f"health-{instance_name}")
    t.start()
    return t
